src/functional/iterator/transforms/unroll_reduce.py

```diff
@@ -1,14 +1,35 @@
+from itertools import groupby
+from typing import Iterable
+
 from eve import NodeTranslator
 from eve.utils import UIDs
 from functional.iterator import ir
 
 
+def all_equal(it: Iterable):
+    g = groupby(it)
+    return next(g, True) and not next(g, False)
+
+
 class UnrollReduce(NodeTranslator):
     @staticmethod
-    def _get_last_offset(node: ir.Expr):
-        assert isinstance(node, ir.FunCall)
-        assert isinstance(node.fun, ir.FunCall) and node.fun.fun == ir.SymRef(id="shift")
-        return node.fun.args[0]
+    def _find_connectivity(reduce_args: Iterable[ir.Expr], offset_provider):
+        connectivities = []
+        for arg in reduce_args:
+            if (
+                isinstance(arg, ir.FunCall)
+                and isinstance(arg.fun, ir.FunCall)
+                and arg.fun.fun == ir.SymRef(id="shift")
+            ):
+                assert isinstance(arg.fun.args[0], ir.OffsetLiteral)
```

Comment:
Is this good enough? I guess there might be cases where we have `shift(dpart)(shift(dfull, 0)(x))` (maybe after some inlining) which after normalization becomes `shift(dfull, 0, dpart)`. So we might actually have to check if the last element is a non-integer offset literal and choose that one?